SwiftCompilerSources/Sources/Optimizer/ModulePasses/DiagnoseUnknownConstValues.swift

```diff
@@ -0,0 +1,128 @@
+//===--------- DiagnoseUnknownConstValues.swift --------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+import SILBridging
+
+struct ConstExprFunctionState {
+  var bridged: BridgedConstExprFunctionState
+
+  init(bridged: BridgedConstExprFunctionState) { self.bridged = bridged }
+  init() {
+    self.bridged = BridgedConstExprFunctionState.create()
+  }
+  
+  mutating func isConstantValue(_ value: Value) -> Bool {
+    return bridged.isConstantValue(value.bridged)
+  }
+  
+  /// TODO: once we have move-only types, make this a real deinit.
+  mutating func deinitialize() {
+    bridged.deinitialize()
+  }
+}
+
+/// Performs mandatory diagnostic pass for emitting errors for '@const' values
+/// which the compiler was not able to simplify/interpret/reduce to a symbolic
+/// value.
+///
+let diagnoseUnknownConstValues = ModulePass(name: "diagnose-unknown-const-values") {
+  (moduleContext: ModulePassContext) in
+  var constExprState: ConstExprFunctionState = .init()
+  defer { constExprState.deinitialize() }
+  
+  // Verify all const globals to be initialized with
+  // compile-time known values
+  verifyGlobals(moduleContext);
+  
+  // Verify @const lets appearing as local variables
+  verifyLocals(moduleContext,
+               constExprState: &constExprState);
+  
+  // For each function call, ensure arguments to @const parameters
+  // are all compile-time known values
+  verifyCallArguments(moduleContext,
+                      constExprState: &constExprState);
+  
+  // For each `@const` function, ensure it is fully
+  // evaluable/interpretable at compile time
+  verifyFunctions(moduleContext);
+}
+
+private func verifyGlobals(_ context: ModulePassContext) {
+  for gv in context.globalVariables.filter({ $0.isConst }) {
```

Comment:
you can also write
```
  for gv in context.globalVariables where gv.isConst {
```

which IMO is a bit easier to read.
And does not create a temporary array (which happens unless you use `.lazy.filter`)!